File: 198-house-robber/198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [None for _ in range(n)]
        dp[n-1] = nums[n-1]
        if n > 1:
            dp[n-2] = max(nums[n-1], nums[n-2])
        for i in range(n-3, -1, -1):
            dp[i] = max(dp[i+2]+nums[i], dp[i+1])
        return dp[0]

File: 198-house-robber/198-house-robber.rs
use std::cmp;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let n: i32 = nums.len() as i32;
        if (n == 1) {
            return nums[0];
        }
        if (n == 2) {
            return cmp::max(nums[0], nums[1]);
        }
        let mut m1: i32 = cmp::max(nums[(n-2) as usize], nums[(n-1) as usize]);
        let mut m2: i32 = nums[(n-1) as usize];
        let mut temp: i32 = 0;
        for i in (0..(n-2)).rev() {
            temp = cmp::max(nums[i as usize] + m2, m1);
            m2 = m1;
            m1 = temp;
        }
        return m1;
    }
}
